### PR TITLE
Allow to pass encoding

### PR DESCRIPTION
**Pull request by Ladas**
So we can send json string as payload and it gets decoded
automatically in a a consumer

### CODE REVIEW

**Comment by Ladas** [comment]
@bzwei @agrare please review and release after merged :-)

File: lib/manageiq/messaging/kafka/common.rb [hunk 1]
```diff
@@ -53,6 +53,7 @@ def queue_for_publish(options)
           body, kafka_opts = for_publish(options)
           kafka_opts[:headers][:message_type] = options[:message] if options[:message]
           kafka_opts[:headers][:class_name] = options[:class_name] if options[:class_name]
+          kafka_opts[:headers][:encoding] = options[:encoding] if options[:encoding]
```
Comment (agrare) [inline lib/manageiq/messaging/kafka/common.rb]: @Ladas will this behavior (not needing to JSON.parse() in the subscriber) be consistent if using artemis?

Comment (Ladas) [reply lib/manageiq/messaging/kafka/common.rb]: ah, probably not, if I am changing this just in kafka?

Comment (agrare) [reply lib/manageiq/messaging/kafka/common.rb]: Yeah so e.g. if the calling code switches to artemis would https://github.com/ManageIQ/topological_inventory-ingress_api/pull/47/files#diff-fdb60c134b290dc6cd1033a065180563R27 break? IDK

Comment (Ladas) [reply lib/manageiq/messaging/kafka/common.rb]: I think we'll need some integration test for this, there might be more issues switching to artemis :-)

so you want to add it also here?
https://github.com/ManageIQ/manageiq-messaging/blob/f93bcf299e13f7e3f89f9562c3f2908a23d854cc/lib/manageiq/messaging/stomp/common.rb#L23

Comment (agrare) [reply lib/manageiq/messaging/kafka/common.rb]: I'm wondering if kafka auto-parses the payload and artemis doesn't we might have to call JSON.parse on the artemis consumer side

Comment (Ladas) [reply lib/manageiq/messaging/kafka/common.rb]: I believe the parsing is on our side in:

https://github.com/ManageIQ/manageiq-messaging/blob/7ce87bccfeab0aef3d3681b1afd8adee9ee241c1/lib/manageiq/messaging/common.rb#L21

Comment (agrare) [reply lib/manageiq/messaging/kafka/common.rb]: Ah okay, then yeah just adding encoding to the headers in stomp/common should do the trick :+1:

Comment (Ladas) [reply lib/manageiq/messaging/kafka/common.rb]: done

**Comment by bzwei** [comment]
@Ladas why https://github.com/ManageIQ/manageiq-messaging/blob/master/lib/manageiq/messaging/client.rb#L14 does not work for you?

**Comment by Ladas** [comment]
@bzwei I have this in https://github.com/ManageIQ/topological_inventory-persister/blob/a69faef4b9ed10da8d5065c9d5e17ce7020be2fe/lib/topological_inventory/persister/worker.rb#L92

But unless I send the :encoding in the header, the `client.subscribe_topic(queue_opts) do |msg|` doesn't parse the json

**Comment by bzwei** [comment]
https://github.com/ManageIQ/manageiq-messaging/blob/master/lib/manageiq/messaging/common.rb#L8 already does what this PR is trying to do. I don't understand what makes the difference.

**Comment by Ladas** [comment]
@bzwei I am sending a string, so that line is not invoked, going through here https://github.com/ManageIQ/manageiq-messaging/blob/master/lib/manageiq/messaging/common.rb#L7, I am sending JSON encoded string

**Comment by bzwei** [comment]
Then don't encode yourself. Send hash as the body, let the gem encode for you. It is asymmetry that you send a JSON string and receive a Hash object.

**Comment by Ladas** [comment]
@bzwei ah, that's it, I have the data already encoded, so I don't want to do `JSON.parse(json_data).to_json` again, it's an expensive operation and I have better things I can do with the CPU (like process more requests :-))

**Comment by Ladas** [comment]
@bzwei @agrare not sure if being asymmetric here is an issue, sending json in and consumer automatically parses the json (when I am manually saying the string is json). That seems fine to me.

**Comment by bzwei** [comment]
Logically it is good to maintain the symmetry. It is internal to the gem how to encode and decode, the application receives the same as what is originally sent. This is similar to many other systems for example the multi-layed internet protocols. In your case if the payload is already encoded by the publisher app, the decoding can be done by the subscriber app.

If we really want to break the symmetry, please document in the comments and README.

**Comment by agrare** [comment]
@bzwei how about we set the encoding _then_ return if `body.kind_of?(String)`?  That way we still allow the client side to automatically decode the message while allowing the producer to pass in already encoded data?

**Comment by bzwei** [comment]
@agrare The current code already works this way as you proposed. Please update the document and/or example. Basically we have the following scenarios:

1. global setting for `encoding`, either `yaml` or `json`.
publish string => no encoding => receive string (up to the subscriber to decide whether and how to decode the string)
publish hash => encode => decode => receive hash

2. message header contains `encoding`, which overrides global setting
publish string => no encoding => decode => receive hash (if the method is either `yaml` or `json`)
publish string => no encoding => receive string (if the method is other type that we can't handle. We should no longer raise an error, but keep the coding method in the header, so that the subscriber knows how to handle. Need a PR)
publish hash => encode => decode => receive hash (Need a PR to use header specific encoding method)

**Comment by Ladas** [comment]
I've just fixed this in https://github.com/ManageIQ/topological_inventory-persister/pull/42/files

**Pull request closed without merging**